sentry-ruby/lib/sentry/profiler.rb

```diff
@@ -0,0 +1,232 @@
+      return nil unless results[:raw]
+
+      frame_map = {}
+
+      frames = results[:frames].to_enum.with_index.map do |frame, idx|
+        frame_id, frame_data = frame
+
+        # need to map over stackprof frame ids to ours
+        frame_map[frame_id] = idx
+
+        in_app = in_app?(frame_data[:file])
+        filename = compute_filename(frame_data[:file], in_app)
+        function, mod = split_module(frame_data[:name])
+
+        {
+          abs_path: frame_data[:file],
+          function: function,
+          module: mod,
+          lineno: frame_data[:line],
+          filename: filename,
+          in_app: in_app
+        }.compact
+      end
+
+      idx = 0
+      stacks = []
+      num_seen = []
+
+      # extract stacks from raw
+      # raw is a single array of [.., len_stack, *stack_frames(len_stack), num_stack_seen , ..]
+      while (len = results[:raw][idx])
+        idx += 1
+
+        # our call graph is reversed
+        stack = results[:raw].slice(idx, len).map { |id| frame_map[id] }.compact.reverse
+        stacks << stack
+
+        num_seen << results[:raw][idx + len]
+        idx += len + 1
+
+        log('Unknown frame in stack') if stack.size != len
+      end
+
+      idx = 0
+      elapsed_since_start_ns = 0
+      samples = []
+
+      num_seen.each_with_index do |n, i|
+        n.times do
+          # stackprof deltas are in microseconds
+          delta = results[:raw_timestamp_deltas][idx]
+          elapsed_since_start_ns += (delta * 1e3).to_i
+          idx += 1
+
+          # Not sure why but some deltas are very small like 0/1 values,
+          # they pollute our flamegraph so just ignore them for now.
+          # Open issue at https://github.com/tmm1/stackprof/issues/201
+          next if delta < 10
+
+          samples << {
+            stack_id: i,
+            # TODO-neel-profiler we need to patch rb_profile_frames and write our own C extension to enable threading info.
+            # Till then, on multi-threaded servers like puma, we will get frames from other active threads when the one
+            # we're profiling is idle/sleeping/waiting for IO etc.
+            # https://bugs.ruby-lang.org/issues/10602
+            thread_id: '0',
+            elapsed_since_start_ns: elapsed_since_start_ns.to_s
+          }
+        end
+      end
+
+      log('Some samples thrown away') if samples.size != results[:samples]
+
+      if samples.size <= 2
+        log('Not enough samples, discarding profiler')
+        return nil
+      end
+
+      profile = {
+        frames: frames,
+        stacks: stacks,
+        samples: samples
+      }
+
+      {
+        event_id: @event_id,
+        platform: PLATFORM,
+        version: VERSION,
+        profile: profile
+      }
+    end
+
+    private
+
+    def log(message)
+      Sentry.logger.debug(LOGGER_PROGNAME) { "[Profiler] #{message}" }
+    end
+
+    def project_root
+      @project_root ||= Sentry.configuration.project_root
```

Comment:
We should be able to define this in the constructor?

Reply:
this is safeguarded by the `Sentry.initialized?` check above otherwise `Sentry.configuration` could be nil. I don't know of a clean way to do this, in `Transaction` for instance, we pass in `hub` manually but that's marked as deprecated and I also don't want to pass in `hub` here.

Reply:
If the SDK is not initialised, it probably won't even create any Transaction (otherwise [these would error](https://github.com/getsentry/sentry-ruby/blob/master/sentry-ruby/lib/sentry/transaction.rb#L76-L82)), which means no `Profiler.new` will be called anyway?

Reply:
yea ok true, then I'll move them up, thanks!

Reply:
this causes a bunch of tests to fail because we don't always initialized sentry in tests, for instance in `hub_spec`...

Reply:
so either we pass in `hub` to `Profiler.new` like with `Transaction` or we need to guard with `Sentry.initialized?` checks like I did. This is more a testing problem but I don't want to change the way we test these things for this PR.

Which do you prefer?

Reply:
I actually prefer not calling `Sentry.configuration` at all from the SDK's key components due to this reason. For example, if we call it inside `Transaction`, then we'll need to initialise the entire SDK to test it, which makes the test hard to maintain and make tests couple.

Since `Transaction` current holds `@configuration`, which was planned to be removed. How about we don't remove it anymore and pass it to `Profiler.new`? This will make testing `Profiler` easier as well.